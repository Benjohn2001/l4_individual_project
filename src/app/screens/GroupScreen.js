import React from 'react';
import { Feather } from '@expo/vector-icons';
import { View, Image, TextInput, Text, TouchableOpacity, SafeAreaView, ScrollView, FlatList, ActivityIndicator } from 'react-native';
import AppButtonPurple from '../components/AppButtonPurple'
import AppButtonLight from '../components/AppButtonLight';
import TwoButtonsSide from '../components/TwoButtonsSide';
import { COLOURS } from '../assets/colours';
import ProfileBox from '../components/ProfileBox';
import RowItem from '../components/RowItem';
import PillButton from '../components/PillButton';
import PressableIcon from '../components/PressableIcon';
import GroupMemberBar from '../components/GroupMemberBar';
import { query, get, ref, getDatabase } from 'firebase/database';
import { auth } from '../../firebase';
import { useIsFocused } from '@react-navigation/native';

const GroupScreen = ({ route, navigation }) => {

    const item=route.params.item
    const name=item.val()["name"]
    const membersRef=item.val()["membersRef"]
    const key=item.key

    const [membersKeys, setMembersKeys] = React.useState([])
    const [members, setMembers] = React.useState([])
    const [showMembers, setShowMembers] = React.useState(false)
    const [fetched, setFetched] = React.useState(false)

    const isFocused= useIsFocused()

    React.useEffect(()=>{
        fetchData()
    },[isFocused, showMembers])

    const fetchData = async () =>{
        setMembersKeys([])
        setMembers([])
        const groupMembersRef = query(ref(getDatabase(), "/groupMembers/"+membersRef))
        const data = await get(groupMembersRef)
        data.forEach(c => {
            setMembersKeys(a => {return [...a , c]})
        })
        for (const i in membersKeys) {
            const userRef = query(ref(getDatabase(), "/users/"+membersKeys[i].val()["member"]))
            const data = await get(userRef)
            setMembers(a => {return [...a , data]})
        }
        setFetched(true)
    }


    return (
        <SafeAreaView className="flex-1  bg-primaryPurple" >
        {fetched ?
            <><View className="flex-row justify-between pt-14 mx-10">
                    <PressableIcon
                        onPress={() => {
                            navigation.goBack();
                        } }
                        icon="arrow-left"
                        size={40}
                        color="black" />
                    <Text className="font-bold ml-auto mr-auto text-3xl">
                        {name}
                    </Text>
                </View>
                <View className="justify-center items-center">
                    <Image
                        className="w-full h-96"
                        source={require("../assets/clock.png")} 
                    />
                </View>
                                    
                    { showMembers ?
                        <>
                        <View className="flex-1">
                            <PillButton
                                title="Hide Members"
                                onPress={() => {
                                    setShowMembers(false);
                                    setMembersKeys([])
                                    setMembers([])
                                } }
                                icon="users" 
                            />
                        
                                <FlatList
                                    showsVerticalScrollIndicator={false}
                                    extraData={members}
                                    data={members}
                                    renderItem={({ item }) => (
                                        <GroupMemberBar
                                            title={item.val()["firstName"] + " " + item.val()["lastName"]}
                                            onPress={() => {
                                                if (item.key == auth.currentUser.uid) {
                                                    navigation.navigate("HomeScreen", { screen: "Me" });
                                                } else {
                                                    navigation.push('UserProfileScreen', {
                                                        user: item,
                                                    });
                                                }
                                            } }
                                            avatar={item.val()["profilePic"]}
                                            color="red" />
                                    )} 
                                />
                            </View>
                        </>
                    :
                        <PillButton
                            title="Show Members"
                            onPress={()=>{
                                setShowMembers(true)
                            }}
                            icon="users"
                        />
                    }
                    
                    <View className="items-center py-3 mt-auto">
                        <TwoButtonsSide
                            title1="Settings"
                            onPress1={() => {
                                navigation.navigate("GroupSettingsScreen",{
                                    item: item,
                                });
                                setShowMembers(false)
                            } }
                            icon1="settings"
                            color1="#E7E7FF"
                            title2="Customise"
                            onPress2={() => {
                                navigation.navigate("ClockCustomiseScreen");
                                setShowMembers(false)
                            } }
                            icon2="clock" 
                            color2="#6B4EFF"/>
                    </View></>
        :
            <View className="justify-center items-center flex-1">
                <ActivityIndicator size="large" color="#6B4EFF"  />
                <Text className="text-center">Loading Group</Text>
            </View>
        }
        </SafeAreaView>

    );
};

export default GroupScreen;