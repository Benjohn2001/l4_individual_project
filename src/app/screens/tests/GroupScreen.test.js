import renderer from "react-test-renderer";
import { render } from "@testing-library/react-native";
import GroupScreen from "../GroupScreen";

jest.mock("firebase/database", () => ({
  getDatabase: jest.fn().mockReturnThis(),
  push: jest.fn().mockReturnThis(),
  ref: jest.fn().mockReturnThis(),
  set: jest.fn().mockReturnThis(),
  query: jest.fn().mockReturnThis(),
  onValue: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/storage", () => ({
  getStorage: jest.fn().mockReturnThis(),
}));

const navigation = {
  navigate: jest.fn(),
  push: jest.fn(),
  goBack: jest.fn(),
};

jest.mock("firebase/app", () => ({
  auth: jest.fn().mockReturnThis(),
  initializeApp: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/auth", () => ({
  auth: jest.fn().mockReturnThis(),
  getAuth: () => ({
    currentUser: jest.fn().mockReturnThis(),
    signOut: () => ({
      then: () => ({
        catch: jest.fn(),
      }),
    }),
  }),
}));

jest.mock("@react-navigation/native", () => ({
    useIsFocused: jest.fn(),
  }));

const route = {
  params: {
    item: {
      val: jest.fn().mockReturnThis(),
    },
  },
};

test("Check GroupScreen loading correctly", async () => {
  render(<GroupScreen navigation={navigation} route={route} />);
  const domTree = renderer
    .create(<GroupScreen route={route} />)
    .toJSON();
  expect(domTree).toMatchSnapshot();
});
