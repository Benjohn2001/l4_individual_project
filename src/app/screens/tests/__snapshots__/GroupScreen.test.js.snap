// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check GroupScreen loading correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      Object {
        "backgroundColor": "#C6C4FF",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "justifyContent": "center",
        },
        Object {
          "alignItems": "center",
        },
        Object {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <ActivityIndicator
      color="#6B4EFF"
      size="large"
    />
    <Text
      style={
        Array [
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      Loading Group
    </Text>
  </View>
</RCTSafeAreaView>
`;
